File: visionTagging/visDatabaseCreate.pl
#! /usr/bin/perl

# cpan DBD::mysql

use warnings;
use strict;

use Cwd;
use DBI;
use Time::localtime;

# use YAML::XS 'LoadFile';
use XML::Simple qw(:strict);

our $base_path = cwd();  # Get the directory of this module. 
$base_path =~ m/(.*)\/.*$/;  # Regex to go up one directory.
$base_path = $1 . "/";  # Capture the output and put it in $base_path.

our $config_file = $base_path . "config/params.xml";
# our $config = LoadFile($YAML_file);  # YAML is more cross-language.


our $config = XMLin( $config_file, ForceArray => 1, KeyAttr=>[]) or die("Sorry! Can't read this XML.");

#TODO: Delete the old database file
#TODO: Database file location better handled

my $visionBase = $config->{'visionTaggingParams'}->[0];

our $dbhandle;
our $dbName = $visionBase->{'database'}->[0]->{'fileName'}->[0];

my $count_args = @ARGV ;

if ($count_args == 1) {
	my $db = $ARGV[0];
	if ($db =~ /db$/){
		print "valid";
		$dbName = $db;
	}
}



dbConnect();
dropTables();
createLoggingTable();
createMetadataTable();
populateMetadataTable();

$dbhandle->disconnect;

# End creation of tables.

sub dbConnect{
	$dbhandle = DBI->connect("dbi:SQLite:$dbName", "user" , "pass");
}

sub dropTables{
	# Clean out the old tables so we can create them afresh.
	my $tableBase = $visionBase->{'database'}->[0]->{'tables'}->[0];
	my $dropPhotos = qq/DROP TABLE IF EXISTS $tableBase->{'recordDataTable'}->[0]->{'Name'}->[0] /;
	my $query = $dbhandle->prepare($dropPhotos);
	$query->execute() or die $DBI::errstr;


	my $dropMetadata = qq/DROP TABLE IF EXISTS $tableBase->{'visionMetaTable'}->[0]->{'Name'}->[0] /;
	$query = $dbhandle->prepare($dropMetadata);
	$query->execute() or die $DBI::errstr;
	
}

###############################################
########### subs for table creation ###########
###############################################

# Create the photo primary key and filename table
sub createLoggingTable{
	# Create the database table, where photo_key is the primary key and photo_file is the file name 
	my $recordBase = $visionBase->{'database'}->[0]->{'tables'}->[0]->{'recordDataTable'}->[0];
	my $sql_quer = qq/CREATE TABLE $recordBase->{'Name'}->[0] (
		    $recordBase->{'Columns'}->[0]->{'File'}->[0] 			STRING,
		    $recordBase->{'Columns'}->[0]->{'Checked'}->[0] 		STRING,
		    $recordBase->{'Columns'}->[0]->{'Portrait'}->[0] 		BOOL,
		    $recordBase->{'Columns'}->[0]->{'Source'}->[0] 			STRING,
		    $recordBase->{'Columns'}->[0]->{'Valid'}->[0]			BOOL
		); /;

	# Prepare and execute the statement
	my $create_handle = $dbhandle->prepare($sql_quer);
	$create_handle->execute() or die $DBI::errstr;
}

sub createMetadataTable{
	my $metaBase = $visionBase->{'database'}->[0]->{'tables'}->[0]->{'visionMetaTable'}->[0];
	my $sql_quer = qq/CREATE TABLE $metaBase->{'Name'}->[0](
		$metaBase->{'Columns'}->[0]->{'NameColumn'}->[0]  STRING,
		$metaBase->{'Columns'}->[0]->{'ValueColumn'}->[0]  STRING
	); /;

	my $create_handle = $dbhandle->prepare($sql_quer);
	$create_handle->execute() or die $DBI::errstr;
}

sub populateMetadataTable{
	my $tm = localtime;
	my $year = $tm->year + 1900;
	my $month = sprintf("%02d", $tm->mon + 1);
	my $day = sprintf("%02d", $tm->mday);
	my $newMonth = "$year-$month-$day";

	my $googFieldsBase = $visionBase->{'database'}->[0]->{'Fields'}->[0]->{'googFields'}->[0];
	my $clarifaiFieldsBase = $visionBase->{'database'}->[0]->{'Fields'}->[0]->{'googFields'}->[0];
	my $metaBase = $visionBase->{'database'}->[0]->{'tables'}->[0]->{'visionMetaTable'}->[0];

	my $populateQuery = qq/INSERT INTO $metaBase->{'Name'}->[0] 
		('$metaBase->{'Columns'}->[0]->{'NameColumn'}->[0]', '$metaBase->{'Columns'}->[0]->{'ValueColumn'}->[0]') VALUES
		('$clarifaiFieldsBase->{'ReadsThisMonth'}->[0]', '0'), 
		('$clarifaiFieldsBase->{'ReadsPerMonth'}->[0]', '5000'),
		('$clarifaiFieldsBase->{'NewMonthDate'}->[0]', '$newMonth'),
		('$clarifaiFieldsBase->{'DayLastRead'}->[0]', '$newMonth'),
		('$clarifaiFieldsBase->{'DayOfNewMonth'}->[0]', '$day'),
		('$googFieldsBase->{'ReadsThisMonth'}->[0]', '0'), 
		('$googFieldsBase->{'ReadsPerMonth'}->[0]', '1000'),
		('$googFieldsBase->{'NewMonthDate'}->[0]', '$newMonth'),
		('$googFieldsBase->{'DayLastRead'}->[0]', '$newMonth'),
		('$googFieldsBase->{'DayOfNewMonth'}->[0]', '$day'); 
		/;

	my $create_handle = $dbhandle->prepare($populateQuery);
	$create_handle->execute() or die $DBI::errstr;

}

# sub update_metadata{

# 	my ($sec,$min,$hour,$mday,$mon,$year) = localtime(time);
# 	$year += 1900;
# 	$mon += 1;

# 	my $dateTime = sprintf("%04d-%02d-%02d %02d:%02d:%02d", $year, $mon, $mday, $hour, $min, $sec);

# 	my $updatedValQuery = qq/INSERT INTO $params::metadataTableName VALUES ("$params::metadataLastEditedField", "$dateTime")/;
	
# 	my $metadata_handle = $dbhandle->prepare($updatedValQuery);
# 	$metadata_handle->execute() or die $DBI::errstr;

# }

File: visionTagging/visDatabaseCreate.py
#! /usr/bin/env python

import sqlite3
from sqlite3 import Error
import os
import xmltodict
import time

project_path = os.path.abspath(os.path.join(__file__,"../.."))
script_path  = os.path.abspath(os.path.join(__file__,".."))

with open(os.path.join(project_path, 'config/params.xml') ) as stream:
    try:
        params = xmltodict.parse(stream.read())
    except Exception as exc:
        print(exc)
        exit(1)



# my $visionBase = $config->{'visionTaggingParams'}->[0];

# our $dbhandle;
# our $dbName = $visionBase->{'database'}->[0]->{'fileName'}->[0];

# my $count_args = @ARGV ;

# if ($count_args == 1) {
#     my $db = $ARGV[0];
#     if ($db =~ /db$/){
#         print "valid";
#         $dbName = $db;
#     }
# }
visionDBParamBase = params['params']['visionTaggingParams']['database']

def connectDB(params):

    visionDBfile = params['params']['visionTaggingParams']['database']['fileName']

    conn = sqlite3.connect(os.path.join(project_path, 'databases', visionDBfile) )
    conn.text_factory = lambda x: unicode(x, "utf-8", "ignore")
    return conn

recordDataTableName = visionDBParamBase['tables']['recordDataTable']['Name']
visionMetaTableName = visionDBParamBase['tables']['visionMetaTable']['Name']

def dropTables(conn):
    # Clean out the old tables so we can create them afresh.
    # tableBase = $visionBase->{'database'}->[0]->{'tables'}->[0];
    dropPhotos = 'DROP TABLE IF EXISTS ' + recordDataTableName

    c = conn.cursor()
    c.execute(dropPhotos)
    conn.commit()


    dropMetadata = 'DROP TABLE IF EXISTS ' + visionMetaTableName
    
    c = conn.cursor()
    c.execute(dropMetadata)
    conn.commit()
    
def createLoggingTable(conn):
    # Create the database table, where photo_key is the primary key and photo_file is the file name 
    recordBase = visionDBParamBase['tables']['recordDataTable']
    logTableCreate = 'CREATE TABLE ' + recordBase['Name'] + ' (' + \
            recordBase['Columns']['File'] + ' STRING, ' + \
            recordBase['Columns']['Checked'] + ' STRING, ' + \
            recordBase['Columns']['Portrait'] + ' BOOL, ' + \
            recordBase['Columns']['Source'] + ' STRING, ' + \
            recordBase['Columns']['Valid'] + ' BOOL); '

    c = conn.cursor()
    c.execute(logTableCreate)
    conn.commit()

def createMetadataTable(conn):
    metaBase = visionDBParamBase['tables']['visionMetaTable']
    metadataCreate = 'CREATE TABLE ' + metaBase['Name'] + ' ( ' + \
        metaBase['Columns']['NameColumn'] + ' STRING, ' + \
        metaBase['Columns']['ValueColumn'] + ' STRING );' 

    c = conn.cursor()
    c.execute(metadataCreate)
    conn.commit()


def populateMetadataTable(conn):

    (year, month, day, hour, minute, second, wday, yearday, isdst) = time.localtime()

    # my $newMonth = "$year-$month-$day";
    newMonth = time.strftime('%Y-%m-%d', time.localtime())

    print newMonth

    googFieldsBase = visionDBParamBase['Fields']['googFields'];
    clarifaiFieldsBase = visionDBParamBase['Fields']['clarifaiFields']
    metaBase = visionDBParamBase['tables']['visionMetaTable']

    populateQuery = 'INSERT INTO ' + metaBase['Name'] + \
        ' ( "' + metaBase['Columns']['NameColumn']  + '" , ' + metaBase['Columns']['ValueColumn'] + ') VALUES ' + \
        ' ( "' + clarifaiFieldsBase['ReadsThisMonth'] + '" , 0),'  + \
        ' ( "' + clarifaiFieldsBase['ReadsPerMonth'] + '" , 5000),' + \
        ' ( "' + clarifaiFieldsBase['NewMonthDate'] + '" , "' + str(newMonth) + '"), ' + \
        ' ( "' + clarifaiFieldsBase['DayLastRead'] + '" , "' + str(newMonth) + '"), ' + \
        ' ( "' + clarifaiFieldsBase['DayOfNewMonth'] + '" , ' + str(day) + '), ' + \
        ' ( "' + googFieldsBase['ReadsThisMonth'] + '" , 0), ' + \
        ' ( "' + googFieldsBase['ReadsPerMonth'] + '" , 10000), ' + \
        ' ( "' + googFieldsBase['NewMonthDate'] + '" , "' + str(newMonth) + '"), ' + \
        ' ( "' + googFieldsBase['DayLastRead'] + '" , "' + str(newMonth) + '"), ' + \
        ' ( "' + googFieldsBase['DayOfNewMonth'] + '" , ' + str(day) + '); '

    print populateQuery

    c = conn.cursor()
    c.execute(populateQuery)
    conn.commit()


conn = connectDB(params)
dropTables(conn)
createLoggingTable(conn)
createMetadataTable(conn)
populateMetadataTable(conn)


# End creation of tables.
